lib/index.js

```diff
@@ -1,350 +1,837 @@
+ * @property {(UserModelCacheConfig | string)[]=} models
+ * @property {Object=} redisConfig
+ */
 
-      // --- Helpers
+/// Resolved types
 
-      /**
-       *
-       * @param {string} model The model used to find related caches to bust
-       * @return {array} Array of related models
-       */
-      const getRelatedModels = (model) => {
-        const models = Object.assign({}, strapi.models, strapi.plugins['users-permissions'].models);
-        const components = strapi.components;
-        const modelKey = _.findKey(models, { 'collectionName': model }) || model;
-        const modelObj = models[modelKey];
-        const attributes = _.get(modelObj, 'allAttributes', {});
-        const relatedComponents = [];
-        const relatedModels = {};
-
-        // first, look for direct relations
-        for (const key in attributes) {
-          const attr = attributes[key];
-
-          if (!attr.via) continue;
-
-          if (attr.collection) {
-            relatedModels[attr.collection] = models[attr.collection];
-          } else if (attr.model && attr.model !== 'file') {
-            relatedModels[attr.model] = models[attr.model];
-          }
+/**
+ * @typedef {Object} ModelCacheConfig
+ * @property {boolean} singleType
+ * @property {boolean} disableHitpass
+ * @property {boolean} injectDefaultRoutes
+ * @property {string[]} headers
+ * @property {string} model
+ * @property {number} maxAge
+ * @property {string=} plugin
+ * @property {CustomRoute[]} routes
+ */
+
+/**
+ * @typedef {Object} MiddlewareCacheConfig
+ * @property {'mem'|'redis'} type
+ * @property {boolean} enabled
+ * @property {boolean} logs
+ * @property {boolean} populateContext
+ * @property {boolean} populateStrapiMiddleware
+ * @property {boolean} enableEtagSupport
+ * @property {boolean} enableXCacheHeaders
+ * @property {boolean} clearRelatedCache
+ * @property {boolean} withKoaContext
+ * @property {boolean} withStrapiMiddleware
+ * @property {number} max
+ * @property {number} maxAge
+ * @property {number} cacheTimeout
+ * @property {string[]} headers
+ * @property {ModelCacheConfig[]} models
+ * @property {Object=} redisConfig
+ */
+
+/**
+ * @param {UserMiddlewareCacheConfig} userOptions
+ * @returns {MiddlewareCacheConfig}
+ */
+const resolveUserOptions = (userOptions) => {
+  /**
+   * @type {MiddlewareCacheConfig}
+   */
+  const defaultOptions = {
+    type: "mem",
+    logs: true,
+    enabled: false,
+    populateContext: false,
+    populateStrapiMiddleware: false,
+    enableEtagSupport: false,
+    enableXCacheHeaders: false,
+    clearRelatedCache: false,
+    withKoaContext: false,
+    withStrapiMiddleware: false,
+    headers: [],
+    max: 500,
+    maxAge: 3600000,
+    cacheTimeout: 500,
+    models: [],
+  };
+
+  const { models = [] } = userOptions;
+  /**
+   * @type {ModelCacheConfig[]}
+   */
+  const resolvedModels = [];
+
+  const defaultModelConfig = {
+    singleType: false,
+    disableHitpass: false,
```

Comment:
I would suggest changing this option from a boolean to 

```typescript
{
  authStrategy: "hitpass" | "none"
}
```

I do hope one day to add proper handling of authenticated requests using the `users-permission` plugin, so that would allow us to simply add a strategy later on

Reply:
hum, how about something like:

```
{
  hitpass: (ctx) => false
}
```

Reply:
resolved ?

Reply:
resolved